_api_app/app/Shared/Storage.php

```diff
@@ -171,7 +178,15 @@ protected function array2xmlFile($arr, $xml_file, $root)
     protected function xmlFile2array($xml_file)
     {
         if (file_exists($xml_file)) {
-            $xml_str = file_get_contents($xml_file);
+
+            $fp = fopen($xml_file, 'r');
+            if (flock($fp, LOCK_EX)) {
```

Comment:
I think we can use a shared lock (`LOCK_SH`) for reading. Meaning multiple processes should be able to read, but the writer will not be able to acquire unique lock until this is released.